core/query/sql.ml

```diff
@@ -137,138 +152,162 @@ let order_by_clause n =
+      (Format.pp_print_list ~pp_sep:pp_comma pp_value) values
+  in
+  match q with
+    | UnionAll ([], _) -> Format.fprintf ppf "%a" pp_constant "select 42 as \"@unit@\" where false"
+    | UnionAll ([q], n) -> Format.fprintf ppf "%a%a" pr_q q pp_constant (order_by_clause n)
+    | UnionAll (qs, n) ->
+      let pp_sep_union ppf () = Format.fprintf ppf " union all " in
+      let pp_value ppf x = Format.fprintf ppf "(%a)" pr_q x in
+      Format.fprintf ppf "%a%a"
+        (Format.pp_print_list ~pp_sep:pp_sep_union pp_value) qs
+        pp_constant (order_by_clause n)
+    | Select (fields, [], Constant (Constant.Bool true), _os) ->
+      Format.fprintf ppf "select %a" pr_fields fields
+    | Select (fields, [], condition, _os) ->
+      Format.fprintf ppf "select * from (select %a) as %a where %a"
+        pr_fields fields
+        pp_constant (fresh_dummy_var ())
+        pr_b condition
+    | Select (fields, tables, condition, os) ->
+        (* using quote_field assumes tables contains table names (not nested queries) *)
+        let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
+        pr_select ppf fields tables condition os
+    | Delete { del_table; del_where } ->
+        pr_delete ppf del_table del_where
+    | Update { upd_table; upd_fields; upd_where } ->
+        pr_update ppf upd_table upd_fields upd_where
+    | Insert { ins_table; ins_fields; ins_records } ->
+        pr_insert ppf ins_table ins_fields ins_records
+    | With (_, q, z, q') ->
+        match q' with
+        | Select (fields, tables, condition, os) ->
+            (* Inline the query *)
+            let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
+            let pr_q ppf q = pr_query ppf quote ignore_fields q in
+            let q = Format.asprintf "(%s) as %s" (Format.asprintf "%a" pr_q q) (string_of_table_var z) in
+            pr_select ppf fields (q::tables) condition os
+        | _ -> assert false
 
-and string_of_base quote one_table b =
-  let sb = string_of_base quote one_table in
+and pr_base ppf quote one_table b =
+  let string_of_projection quote one_table (var, label) =
+  if one_table then
+    quote label
+  else
+    buffer_concat [string_of_table_var var; "."; (quote label)]
+  in
+  let pr_b ppf b = pr_base ppf quote one_table b in
+  let pr_q ppf q = pr_query ppf quote true q in
+  let unary_ops = StringSet.of_list ["intToString"; "stringToInt"; "intToFloat"; "floatToString"; "stringToFloat"; "floatToInt"; "not"; "negate"; "negatef"] in
+  let binary_ops = StringSet.of_list ["&&"; "||"; "=="; "<>"; "<"; ">"; "<="; ">="; "RLIKE"; "LIKE"] in
+  let binary_map op =
+    match op with
+      | "&&" -> "and"
+      | "||" -> "or"
+      | "==" -> "="
+      | _ -> op
+  in
+  let unary_map op =
+    match op with
+      | "floatToInt" -> "floor"
+      | "not" -> "not "
+      | "negate" | "negatef" -> "-"
+      | _ -> ""
```

Comment:
This is because the other operations are type conversions, but looks like a potential bug.  (and is fragile: it depends on the fact that all the operations aside from floatToint, not and negate correspond to silent type conversions in SQL, and could become a bug if we add other unary operations for which this isn't the case).  Could you add a comment or make the pattern match explicit so that it's more obvious that "" is an appropriate translation for the remaining operations?